browser-specific-failures.js

```diff
@@ -0,0 +1,386 @@
+      for (const [browser, count] of counts) {
+        const score = count / denominator;
+        scores.set(browser, scores.get(browser) + score);
+      }
+    }
+  }
+
+  return scores;
+}
+
+// See documentation of advanceDateToSkipBadDataIfNecessary. These ranges are
+// inclusive, exclusive.
+const STABLE_BAD_RANGES = [
+    // This was some form of Safari outage, undiagnosed but a clear erroneous
+    // spike in failure rates.
+    [ moment("2019-02-06"), moment("2019-03-04") ],
+    // This was a Safaridriver outage, resolved by
+    // https://github.com/web-platform-tests/wpt/pull/18585
+    [ moment("2019-06-27"), moment("2019-08-23") ],
+];
+const EXPERIMENTAL_BAD_RANGES = [
+    // This was a Safaridriver outage, resolved by
+    // https://github.com/web-platform-tests/wpt/pull/18585
+    [ moment("2019-06-27"), moment("2019-08-23") ],
+];
+
+// There have been periods where results cannot be considered valid and
+// contribute noise to the metrics. Given a date, this function advances it as
+// necessary to avoid bad data.
+//
+// TODO(smcgruer): Take into account --products being used.
+function advanceDateToSkipBadDataIfNecessary(date, experimental) {
+  const ranges = experimental ? EXPERIMENTAL_BAD_RANGES : STABLE_BAD_RANGES;
+  for (const range of ranges) {
+    if (date >= range[0] && date < range[1]) {
+      console.log(`Skipping from ${date.format('YYYY-MM-DD')} to ${range[1].format('YYYY-MM-DD')} due to bad data`);
+      return range[1];
+    }
+  }
+  return date;
+}
+
+const RUNS_URI = 'https://wpt.fyi/api/runs?aligned=true&max-count=1';
+
+// Fetches aligned runs from the wpt.fyi server, between the |from| and |to|
+// dates. If |experimental| is true fetch experimental runs, else stable runs.
+// Returns a map of date to list of runs for that date (one per product)
+//
+// TODO: Known problem: there are periods of time, mostly mid-late 2018, where
+// we ran both Safari 11.1 and 12.1, and the results are massively different.
+// We should fetch multiple runs for each browser and have upgrade logic.
+async function fetchAlignedRunsFromServer(products, from, to, experimental) {
+  const label = experimental ? 'experimental' : 'stable';
+  let params = `&labels=master,${label}`;
+  for (const product of products) {
+    params += `&product=${product}`;
+  }
+  const runsUri = `${RUNS_URI}${params}`;
+
+  console.log(`Fetching aligned runs from ${from.format('YYYY-MM-DD')} to ${to.format('YYYY-MM-DD')}`);
+
+  let cachedCount = 0;
+  let before = moment();
+  const alignedRuns = new Map();
+  while (from < to) {
+    const formatted_from = from.format('YYYY-MM-DD');
+    from.add(1, 'days');
+    const formatted_to = from.format('YYYY-MM-DD');
+
+    // We advance the date (if necessary) before doing anything more, so that
+    // code later in the loop body can just 'continue' without checking.
+    from = advanceDateToSkipBadDataIfNecessary(from, experimental);
+
+    // Attempt to read the runs from the cache.
```

Comment:
I haven't looked at the implementation, but https://github.com/tidoust/fetch-filecache-for-crawling *might* be useful here. This works fine though, I presume.